### ISSUE TITLE
Built-in way to paste emojis doesn't work

### ISSUE THREAD

**Issue by shadow-identity** [PowerShell/PowerShell#15256]
Emojis do not work either in the input line or in the applications output. This cannot be done even using the standard Windows built-in tool. This feature has been in all competitive shells for years.

## Steps to reproduce

* Use Win + . to open the Windows Emoji Keyboard
* Select any emoji to insert

## Expected behavior

the selected emoji must be inserted

## Actual behavior

Wrong symbol inserted instead
![image](https://user-images.githubusercontent.com/1223112/115105613-5fed5c00-9f60-11eb-97d1-df842b6daf01.png)

## Environment data
Win 10 pro 20H2

```none
Name                           Value
----                           -----
PSVersion                      7.1.3
PSEdition                      Core
GitCommitId                    7.1.3
OS                             Microsoft Windows 10.0.19042
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by jborean93** [comment]
This issue isn’t a problem with PowerShell or any other shells in Windows but rather the console host/terminal you are using. The conhost application is started up when you start PowerShell and it has very limited support for Unicode characters. This has nothing to do with limitations in PowerShell.

What MS are recommending you to use is the Windows Terminal app which has full support for Unicode characters including emojis and the simple paste into the terminal should work. The downside is right now you cannot say use Windows Terminal when you run pwsh, you need to start pwsh from Windows Terminal which is more how things work on non-Windows hosts.

**Comment by rkeithhill** [comment]
From Windows Terminal pasting via the WIndows Emoji selector works when running WSL (Ubuntu 18.04) but it does NOT work when running PowerShell 7.1.3 (or CMD).  So perhaps an issue with conhost which both PS and CMD use but Ubuntu does not.  This should probably be filed on the `Windows Terminal` repo (conhost impl lives there).  Actually, first checkout these existing issues on that repo:

https://github.com/microsoft/terminal/issues/190
https://github.com/microsoft/terminal/issues/1503

**Comment by jborean93** [comment]
Ah my apologies I thought it was possible in WT. It does seem like PSReadLine might have issues reading surrogate pair UTF-16 characters. When trying to input 😊 into Windows Terminal I get the following traceback from PSReadLine

```
Oops, something went wrong.
Please report this bug with ALL the details below, including both the 'Environment' and 'Exception' sections.
Please report on GitHub: https://github.com/PowerShell/PSReadLine/issues/new?template=Bug_Report.md
Thank you!

### Environment
PSReadLine: 2.1.0
PowerShell: 7.1.3
OS: Microsoft Windows 10.0.20334
BufferWidth: 120
BufferHeight: 30

Last 2 Keys
 � Enter

### Exception
System.Text.EncoderFallbackException: Unable to translate Unicode character \\uD83D at index 0 to specified code page.
   at System.Text.EncoderExceptionFallbackBuffer.Fallback(Char charUnknown, Int32 index)
   at System.Text.EncoderFallbackBuffer.InternalFallback(ReadOnlySpan`1 chars, Int32& charsConsumed)
   at System.Text.Encoding.GetBytesWithFallback(ReadOnlySpan`1 chars, Int32 originalCharsLength, Span`1 bytes, Int32 originalBytesLength, EncoderNLS encoder)
   at System.Text.Encoding.GetBytesWithFallback(Char* pOriginalChars, Int32 originalCharCount, Byte* pOriginalBytes, Int32 originalByteCount, Int32 charsConsumedSoFar, Int32 bytesWrittenSoFar, EncoderNLS encoder)
   at System.Text.Encoding.GetBytes(Char* pChars, Int32 charCount, Byte* pBytes, Int32 byteCount, EncoderNLS encoder)
   at System.Text.EncoderNLS.GetBytes(Char* chars, Int32 charCount, Byte* bytes, Int32 byteCount, Boolean flush)
   at System.Text.EncoderNLS.GetBytes(Char[] chars, Int32 charIndex, Int32 charCount, Byte[] bytes, Int32 byteIndex, Boolean flush)
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.IO.StreamWriter.Dispose(Boolean disposing)
   at System.IO.TextWriter.Dispose()
   at Microsoft.PowerShell.PSConsoleReadLine.<>c__DisplayClass83_0.<WriteHistoryRange>b__0()
   at Microsoft.PowerShell.PSConsoleReadLine.WithHistoryFileMutexDo(Int32 timeout, Action action)
   at Microsoft.PowerShell.PSConsoleReadLine.WriteHistoryRange(Int32 start, Int32 end, Boolean overwritten)
   at Microsoft.PowerShell.PSConsoleReadLine.IncrementalHistoryWrite()
   at Microsoft.PowerShell.PSConsoleReadLine.MaybeAddToHistory(String result, List`1 edits, Int32 undoEditIndex, Boolean fromDifferentSession, Boolean fromInitialRead)
   at Microsoft.PowerShell.PSConsoleReadLine.InputLoop()
   at Microsoft.PowerShell.PSConsoleReadLine.ReadLine(Runspace runspace, EngineIntrinsics engineIntrinsics, CancellationToken cancellationToken)
```

The character when UTF-16 encoded is a surrogate pair (4 bytes) that's encoded as `0xD83D 0xDE0A`. You can see in the exception it tried to only encode the first part of the pair which by itself isn't a valid character. It does seem like more work needs to be done in Windows Terminal to actually display those chars on input but maybe PSReadLine needs some extra logic to handle surrogate points.

I also could totally be wrong here so happy for anyone to correct me.

**Comment by jborean93** [comment]
Looks like there is an open issue for this as well https://github.com/PowerShell/PSReadLine/issues/2358.

**Comment by ninmonkey** [comment]
from that thread, it sounds like it is sending one `code-unit` (2bytes) at a time, rather than the `4byte` codepoint. If true that makes sense why there's an exception when I'm using `utf8`. However, it breaks on `unicode` (ie: `utf16le`) too.  

```ps1
    # testing ⏰                                            
    while ($true) {                                         
        '0x{0:X4}' -f [int][Console]::ReadKey($true).KeyChar
    }                                                       
0x23F0                                                      
                                                            
    # testing 🐒                                            
    while ($true) {                                         
        '0x{0:X4}' -f [int][Console]::ReadKey($true).KeyChar
    }                                                       
0xD83D                                                      
0xDC12                                                      
```

## Breaks on either encoding

For both tests I:

1. Started a new session
2. Disabled PSReadlineSuggestions
3. Set all 3 encodings to the same type
4. type `'` then `win+.`

## Test 1: UTF-16le

<details><summary>
 (Click for full log)

```ps1
$OutputEncoding = [console]::InputEncoding = [console]::OutputEncoding = [System.Text.UnicodeEncoding]::new()
```

</summary>

```ps1
🐒> Set-PSReadLineOption -PredictionSource None
$OutputEncoding = [console]::InputEncoding = [console]::OutputEncoding = [System.Text.UnicodeEncoding]::new()
Get-ConsoleEncoding

Name                      Encoding CodePage isSingleByte
----                      -------- -------- ------------
OutputEncoding            Unicode      1200        False
[console]::InputEncoding  Unicode      1200        False
[console]::OutputEncoding Unicode      1200        False


🐒> # now paste a '😁' smiley using "win+." hotkey

🐒> '�'

Oops, something went wrong.
Please report this bug with ALL the details below, including both the 'Environment' and 'Exception' sections.
Please report on GitHub: https://github.com/PowerShell/PSReadLine/issues/new?template=Bug_Report.md
Thank you!

### Environment
PSReadLine: 2.2.0-beta2
PowerShell: 7.1.2
OS: Microsoft Windows 10.0.19041
BufferWidth: 213
BufferHeight: 71

Last 67 Keys

Ctrl+l Ctrl+v Home Delete Delete [ Enter
UpArrow UpArrow Enter
Backspace UpArrow UpArrow Alt+Enter g e t - c o n s o l e e n Tab Enter
UpArrow Ctrl+a Ctrl+x UpArrow Home UpArrow $ O u t p u t E n c o d i n g Spacebar = Spacebar Enter
UpArrow Ctrl+a Ctrl+x Ctrl+v Enter
UpArrow UpArrow UpArrow Enter
' � ' Enter

### Exception

System.Text.EncoderFallbackException: Unable to translate Unicode character \\uD83D at index 1 to specified code page.
at System.Text.EncoderExceptionFallbackBuffer.Fallback(Char charUnknown, Int32 index)
at System.Text.EncoderFallbackBuffer.InternalFallback(ReadOnlySpan`1 chars, Int32& charsConsumed)
at System.Text.Encoding.GetBytesWithFallback(ReadOnlySpan`1 chars, Int32 originalCharsLength, Span`1 bytes, Int32 originalBytesLength, EncoderNLS encoder)
at System.Text.Encoding.GetBytesWithFallback(Char* pOriginalChars, Int32 originalCharCount, Byte* pOriginalBytes, Int32 originalByteCount, Int32 charsConsumedSoFar, Int32 bytesWrittenSoFar, EncoderNLS encoder)
at System.Text.Encoding.GetBytes(Char* pChars, Int32 charCount, Byte* pBytes, Int32 byteCount, EncoderNLS encoder)
at System.Text.EncoderNLS.GetBytes(Char[] chars, Int32 charIndex, Int32 charCount, Byte[] bytes, Int32 byteIndex, Boolean flush)
at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
at System.IO.StreamWriter.Dispose(Boolean disposing)
at System.IO.TextWriter.Dispose()
at Microsoft.PowerShell.PSConsoleReadLine.<>c__DisplayClass98_0.<WriteHistoryRange>b__0()
at Microsoft.PowerShell.PSConsoleReadLine.WithHistoryFileMutexDo(Int32 timeout, Action action)
at Microsoft.PowerShell.PSConsoleReadLine.WriteHistoryRange(Int32 start, Int32 end, Boolean overwritten)
at Microsoft.PowerShell.PSConsoleReadLine.IncrementalHistoryWrite()
at Microsoft.PowerShell.PSConsoleReadLine.MaybeAddToHistory(String result, List`1 edits, Int32 undoEditIndex, Boolean fromDifferentSession, Boolean fromInitialRead)
at Microsoft.PowerShell.PSConsoleReadLine.InputLoop()
at Microsoft.PowerShell.PSConsoleReadLine.ReadLine(Get-Runspace runspace, EngineIntrinsics engineIntrinsics, CancellationToken cancellationToken)

🐒> '�'
�

🐒> Get-ConsoleEncoding

Name                      Encoding        CodePage isSingleByte
----                      --------        -------- ------------
OutputEncoding            Unicode             1200        False
[console]::InputEncoding  Unicode             1200        False
[console]::OutputEncoding Unicode (UTF-8)    65001        False
```
</details>

## Test 1: UTF-8

<details><summary>
 (Click for full log)

```ps1
$OutputEncoding = [console]::InputEncoding = [console]::OutputEncoding = [System.Text.UTF8Encoding]::new()
```

</summary>

```ps1
🐒> Set-PSReadLineOption -PredictionSource None
$OutputEncoding = [console]::InputEncoding = [console]::OutputEncoding = [System.Text.UTF8Encoding]::new()
Get-ConsoleEncoding

Name                      Encoding        CodePage isSingleByte
----                      --------        -------- ------------
OutputEncoding            Unicode (UTF-8)    65001        False
[console]::InputEncoding  Unicode (UTF-8)    65001        False
[console]::OutputEncoding Unicode (UTF-8)    65001        False

🐒> # now paste a '😁' smiley using "win+." hotkey

🐒> '�'

Oops, something went wrong.
Please report this bug with ALL the details below, including both the 'Environment' and 'Exception' sections.
Please report on GitHub: https://github.com/PowerShell/PSReadLine/issues/new?template=Bug_Report.md
Thank you!

### Environment
PSReadLine: 2.2.0-beta2
PowerShell: 7.1.2
OS: Microsoft Windows 10.0.19041
BufferWidth: 213
BufferHeight: 84

Last 31 Keys

Ctrl+v Home Home Delete Delete Delete Delete End Home DownArrow Shift+End Ctrl+v Ctrl+z Ctrl+v Ctrl+z Shift+Home Ctrl+v Home $ Home End DownArrow Enter
UpArrow UpArrow UpArrow Enter
' � ' Enter

### Exception

System.Text.EncoderFallbackException: Unable to translate Unicode character \\uD83D at index 1 to specified code page.
at System.Text.EncoderExceptionFallbackBuffer.Fallback(Char charUnknown, Int32 index)
at System.Text.EncoderFallbackBuffer.InternalFallback(ReadOnlySpan`1 chars, Int32& charsConsumed)
at System.Text.Encoding.GetBytesWithFallback(ReadOnlySpan`1 chars, Int32 originalCharsLength, Span`1 bytes, Int32 originalBytesLength, EncoderNLS encoder)
at System.Text.Encoding.GetBytesWithFallback(Char* pOriginalChars, Int32 originalCharCount, Byte* pOriginalBytes, Int32 originalByteCount, Int32 charsConsumedSoFar, Int32 bytesWrittenSoFar, EncoderNLS encoder)
at System.Text.Encoding.GetBytes(Char* pChars, Int32 charCount, Byte* pBytes, Int32 byteCount, EncoderNLS encoder)
at System.Text.EncoderNLS.GetBytes(Char[] chars, Int32 charIndex, Int32 charCount, Byte[] bytes, Int32 byteIndex, Boolean flush)
at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
at System.IO.StreamWriter.Dispose(Boolean disposing)
at System.IO.TextWriter.Dispose()
at Microsoft.PowerShell.PSConsoleReadLine.<>c__DisplayClass98_0.<WriteHistoryRange>b__0()
at Microsoft.PowerShell.PSConsoleReadLine.WithHistoryFileMutexDo(Int32 timeout, Action action)
at Microsoft.PowerShell.PSConsoleReadLine.WriteHistoryRange(Int32 start, Int32 end, Boolean overwritten)
at Microsoft.PowerShell.PSConsoleReadLine.IncrementalHistoryWrite()
at Microsoft.PowerShell.PSConsoleReadLine.MaybeAddToHistory(String result, List`1 edits, Int32 undoEditIndex, Boolean fromDifferentSession, Boolean fromInitialRead)
at Microsoft.PowerShell.PSConsoleReadLine.InputLoop()
at Microsoft.PowerShell.PSConsoleReadLine.ReadLine(Get-Runspace runspace, EngineIntrinsics engineIntrinsics, CancellationToken cancellationToken)

🐒> '�'
�

🐒> Get-ConsoleEncoding

Name                      Encoding        CodePage isSingleByte
----                      --------        -------- ------------
OutputEncoding            Unicode (UTF-8)    65001        False
[console]::InputEncoding  Unicode (UTF-8)    65001        False
[console]::OutputEncoding Unicode (UTF-8)    65001        False
```

</details>

## Interesting

@jborean93 :

- When testing as `utf16le` and it crashes, it changes `[console]::OutputEncoding` to `utf8` by itself.
- It left `[console]::InputEncoding` and `$OutputEncoding` set as `utf16le`
- it **breaks** when `Utf16SequenceLength > 1`
- it **works** when `Utf16SequenceLength == 1`

That's why the other thread had no problem with `⏰`, but `😃` would crash
```ps1
🐒> '😃⏰'.EnumerateRunes() | ft                                 
                                                                 
IsAscii IsBmp Plane Utf16SequenceLength Utf8SequenceLength  Value
------- ----- ----- ------------------- ------------------  -----
  False False     1                   2                  4 128515
  False  True     0                   1                  3   9200
```

**Comment by jborean93** [comment]
So there's definitely a problem with PSReadLine here, I can type in `Remove-Module -Name PSReadLine` then paste in emojiis without a failure. The pasted value is still not displayed properly and seems to be treated as 2 separate chars (because it's a surrogate pair) but the raw byte value is read correctly by PowerShell. For this test I am using `😃` as referenced in @ninmonkey post just above. This value as UTF-8 encoded is `0xF0 0x9F 0x98 0x83`.

![image](https://user-images.githubusercontent.com/8462645/115135564-cb1c5880-a05c-11eb-8f1b-62914498e451.png)

So you can paste surrogate pair chars into Windows Terminal and it is read as a correct value in pwsh as long as PSReadLine isn't involved but the text that is printed to represent that char isn't correct. Windows Terminal does support outputting those chars due to the ConPTY work that it has implemented.

TLDR I think there are 2 bugs

* PSReadLine is unable to handle unicode characters represented by surrogate pair in UTF-16 causing the stack trace failure
  * Issues here https://github.com/PowerShell/PowerShell/issues/15256#issuecomment-821903005
* Windows Terminal is unable to display those initial input characters as they are typed/pasted into the terminal
  * Maybe https://github.com/microsoft/terminal/issues/190 and https://github.com/microsoft/terminal/issues/1503
  * I'm less confident about these problems as WT does support emojis on output as seen by the screenshot
  * Someone more knowledgeable on this can probably provide better info here

**Comment by ninmonkey** [comment]
In regards to **Windows Terminal**, I am now able to paste, and use the `start+.` hotkey to insert glyphs.

![image](https://user-images.githubusercontent.com/3892031/144862148-7f68f335-2b72-424a-baaf-b8424462e395.png)

I tested `5.1`, `7.1`, and `7.2` -- they all worked on  `😃`

**Comment by rkeithhill** [comment]
Although backspace only deletes half of the emoji i.e., it takes two backspaces to entirely remove the emoji.

**Comment by ninmonkey** [comment]
> Although backspace only deletes half of the emoji i.e., it takes two backspaces to entirely remove the emoji.

That makes sense because when encoded as utf16, `😃` [has a sequence length of 2](https://docs.microsoft.com/en-us/dotnet/api/system.text.rune.utf16sequencelength?view=net-5.0#System_Text_Rune_Utf16SequenceLength) (2 code-units required to represent one codepoint)

![image](https://user-images.githubusercontent.com/3892031/145143312-e1f2233e-38dc-4399-9028-f53a6e714484.png)

```ps1
$grapheme = "`u{1f469}`u{1f3fc}`u{200d}`u{1f9b3}"
$grapheme.EnumerateRunes() | ft -AutoSize
```

[the grapheme](https://manishearth.github.io/blog/2017/01/14/stop-ascribing-meaning-to-unicode-code-points/#grapheme-clusters) `👩🏼‍🦳` [woman-medium-light-skin-tone-white-hair](https://emojipedia.org/woman-medium-light-skin-tone-white-hair/) requires you to hit backspace 7 times

![image](https://user-images.githubusercontent.com/3892031/145144114-8236c56c-16ca-4c72-9871-67a4002f407c.png)

What is a 'single character' in Unicode is complicated. You can see on my screenshot, it doesn't even render `woman-medium-light-skin-tone-white-hair` as one "character". Depending on what version of Unicode your system supports, different people will see different number of "characters" for the exact same string.

![image](https://user-images.githubusercontent.com/3892031/145144567-6d63a7bc-5379-4168-818c-84a8218aaf05.png)
